tools/test-runner/src/buck_test.rs

```diff
@@ -0,0 +1,253 @@
+use std::collections::{HashMap, HashSet};
+use std::env;
+use std::fs::File;
+use std::io::{BufReader, Read};
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+use std::time::{Duration, Instant};
+
+use anyhow::{bail, Context, Result};
+use serde::Deserialize;
+
+use super::pyunit;
+use super::rust;
+
+/// Unit test.
+#[derive(Debug)]
+pub struct Test {
+    /// The command used to execute the test.
+    pub command: Command,
+
+    /// User-facing identifier for this specific test.
+    pub name: String,
+
+    /// Labels/tags associated to this test.
+    pub labels: HashSet<String>,
+
+    /// Contacts for further information.
+    pub contacts: HashSet<String>,
+
+    /// Which type of test this is.
+    pub kind: TestKind,
+}
+
+/// Supported types of tests.
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum TestKind {
+    /// Buck's custom `unittest` runner.
+    Pyunit,
+
+    /// Buck's custom Rust tester.
+    Rust,
+
+    /// Simply runs a command and checks for non-zero exit code.
+    /// When running a `(buck_)sh_test`, this tag is automatically inserted.
+    #[serde(rename = "custom")]
+    Shell,
+}
+
+/// Labels which mark buck test targets for automatic (and silent) exclusion.
+const EXCLUDED_LABELS: &[&str] = &["disabled", "exclude_test_if_transitive_dep"];
+
+#[derive(Debug)]
+pub struct TestResult {
+    pub name: String,
+    pub attempts: u32,
+    pub passed: bool,
+    pub duration: Duration,
+    pub stdout: String,
+    pub stderr: String,
+    pub contacts: HashSet<String>,
+}
+
+pub fn run(mut test: Test, retries: u32) -> TestResult {
```

Comment:
I think it would be nice to put this in an `impl` block of `Test` itself, instead of accepting a mutable `Test` as a parameter

Reply:
Oh I didn't know you could make a standalone `impl` block (without specifying a trait). Just pushed a commit with this, but the reference to the test still needs to be mutable since `command.spawn()` mutates.